cordova-lib/src/cordova/platform.js

```diff
@@ -122,6 +122,12 @@ function addHelper(cmd, hooksRunner, projectRoot, targets, opts) {
                         if (platformAlreadyAdded) {
                             throw new CordovaError('Platform ' + platform + ' already added.');
                         }
+
+                        // Remove the <platform>.json file from the plugins directory, so we start clean (otherwise we
+                        // can get into trouble not installing plugins if someone deletes the platform folder but
+                        // <platform>.json still exists).
+                        removePlatformPluginsJson(projectRoot, target);
```

Comment:
Regarding your first point: If you look at the context of where the call was added, it is within an `if (cmd == 'add')` check.

Regarding your second point - no, we still want the file deleted even if the `add` operation fails. Essentially we're just putting things into the state they would have been in if the platform had been properly removed, so we want to get into that state as early as possible.